### DISCUSSION TITLE
Winforms ImageAttributes equivalent in Avalonia

### DISCUSSION THREAD

**Discussion by chris835** [AvaloniaUI/Avalonia#14015] [Q&A] [answered]
In Winforms, there is an ImageAttributes class which can be used when drawing images. One of the features of that allows one to specify a color matrix to tint or otherwise screw about with the colors of an image.

I believe WPF has a way of achieving the same effect.

Is there a convenient way to do this in Avalonia? The obvious way would be a custom pixel shader, but I'm not sure if that's the way to go in Avalonia.

**Comment by thevortexcloud** [comment]
I don't think Avalonia itself can do this. But you can do exactly this with Skia, since Avalonia is built around Skia you can use it to generate the bitmap without any extra dependencies and then load the result into a `Bitmap` to display it. There is an example here for Xamarin on how to do it. The general steps should be the same for Avalonia.

https://learn.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/effects/color-filters

**Reply by chris835** [reply]
Ah, nice. Skia itself seems to have what I need and beyond.

Ideally, I'd like to not generate another bitmap for rendering but instead just customize the rendering, which looks like it should be an easy thing for Skia to do. So I've been playing around with that here a bit tonight and pretty quickly ran into issues with going from avalonias Bitmap to SKBitmap. When using the Skia backend, the platform implementation of Bitmap seems to be a SKBitmap (wrapped in an ImmutableBitmap), but I can't find a good way to get access to that since all the implementation is internal. This of course makes sense in most cases as the user shouldn't know what's going on internally. Bit of a bummer when there's missing functionality though.

Any ideas on that perhaps? Hopefully I'm missing something.

**Reply by thevortexcloud** [reply, marked as answer]
As far as I am aware, there is no direct way to modify Avalonia's bitmap implementation. It's supposed to be just an abstract wrapper around Skia/what ever else. Instead what you can do is implement your own Bitmap that just happens to also be based on Skia. There is some comments in this discussion here showing how to do that and should also give you direct access to the canvas Avalonia uses.

https://github.com/AvaloniaUI/Avalonia/discussions/13610#discussioncomment-7576208

EDIT: There also the somewhat more esoteric `WriteableBitmap` which from my understanding lets you draw raw pixel data to a bitmap. I however have never used it. Avalonia does not have a lot of documentation on using it. But I am guessing it's based on the `Windows.Media.Imaging.WriteableBitmap`, so it's probably fairly similar to use.

https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.imaging.writeablebitmap?view=windowsdesktop-8.0

**Reply by chris835** [reply]
Ah, I did come across that discussion earlier but didn't think I needed to go that far. I just did a test run and it does the job. Feels a bit weird that the bitmap itself is now pretty much responsible for doing the color tinting, but hey, if it works it works. Thanks for the pointers!

`WriteableBitmap` would absolutely work, and I've used that for other things (such as running Doom in Avalonia), and it works great. But it would need more memory + processing on the CPU. I assume doing the color effect with Skia runs it on the GPU directly, which is nice.

Now the only thing remaining will be to figure out where the code responsible for handling the source property in `<Image Source="/assets/foo.bar"/>` is so I can screw with it to instead load the image using my custom bitmap implementation.

**Reply by thevortexcloud** [reply]
>  Thanks for the pointers!

You are welcome!

> Now the only thing remaining will be to figure out where the code responsible for handling the source property in <Image Source="/assets/foo.bar"/> is so I can screw with it to instead load the image using my custom bitmap implementation.


It's here

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Avalonia.Controls/Image.cs#L15-L19

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Avalonia.Controls/Image.cs#L46-L50
It looks like it gets used in this render method:

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Avalonia.Controls/Image.cs#L73-L96

What it does from there however I can't say.


Interestingly the property apparently belongs to `IImage` rather than `Image`. So I guess you just need to read the value in your implementation?

EDIT: Actually, it looks like `IImage` does not have a Source property so I am not clear why it's set up this way.

**Reply by chris835** [reply]
Yeah I've been looking around that file too..! The part I haven't found yet is whichever part takes the string `"assets/foo.bar"` from the XAML and instantiates and loads the correct bitmap/image platform implementation. The property takes an `IImage` but can be input as a string... so there must be some code somewhere that deals with that. But I haven't been looking around the sources for Avalonia long enough to know where to look..!

**Reply by chris835** [reply]
Ah, here we go:

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguage.cs#L125-L126

**Reply by thevortexcloud** [reply]
I would guess there is some sort of implicit converter somewhere. Possibly this:

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Markup/Avalonia.Markup.Xaml/Converters/BitmapTypeConverter.cs

Assuming that is it, then I don't think there is much you can do since it's hardcoded to `Bitmap`.

 However I don't think I have actually every loaded an image like that in Avalonia. I typically just bind to my bitmaps directly.

**Reply by chris835** [reply]
Yeah, maybe there's a way to override it, but it would probably break other things. I'm probably going to end up just creating a custom `Image` control that has the extra features that I need.

Btw, how did you get that code to embed so nicely above? Couldn't get the link I added to do the same...

**Reply by thevortexcloud** [reply]
> Btw, how did you get that code to embed so nicely above? Couldn't get the link I added to do the same...

It's very finicky. I have to manually fix it 90% of the time (or sometimes just build the URL by hand). But you can't specify columns in the link, only line numbers. In your case it would be

https://github.com/AvaloniaUI/Avalonia/blob/75e7c8d0207c514b0864440d7c918c199be7baea/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguage.cs#L125-L126

**Reply by chris835** [reply]
Ah, good to know! And thanks again!
